Add Product Query Support for Atomic Rating Block

danielwrobert opened the pull request:
Adds Product Query support for the atomic Rating block.

On the client side, when the Rating Block is used within the Product Query block, the markup will be rendered on the server side - no javascript related to the Rating block will be loaded.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #7330 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


### Testing

1. Checkout this branch and run `npm start` (need to run the `start` command to catch the experimental flags).
1. Be sure that you have the latest version of the Gutenberg feature plugin installed and activated.
1. Add the `woocommerce/product-rating` block to the Product Query `INNER_BLOCKS_TEMPLATE` in `assets/js/blocks/product-query/constants.ts` under line 54 (**only for testing, remove afterwards**).
1. Choose at least one product and add a rating to it (this can be done on a single product view for any product).
1. Create a post/page and add the Product Query block.
1. Save it.
1. Open the frontend side and confirm that:
    - No JS related to the Product Rating block is loaded (search in your network tab for `product-rating-frontend.js`).
    - The UI of the block is identical to on the editor side.
    - That the settings are reflected correctly (eg: custom style).
1. Confirm that the rating stars for the product that you've added the rating to appear as expected in the UI.
1. Back in the editor confirm that the Product Rating block can be added as a child of the Product query block via the inserter.
1. For comparison, switch to trunk and add the All Products block to confirm that the aforementioned `product-rating-frontend.js` _is_ loaded.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

danielwrobert committed:
Add PQ support for client-side.

Set up the block for PQ support and add necessary adjustments for the
editor. Will address dynamic save functionality in a following commit.

danielwrobert committed:
Add dynamic render function for PQ support.

danielwrobert committed:
Add dynamic render callback for SSR.

danielwrobert committed:
Remove client-side Save function.

danielwrobert commented:
~~I forgot about converting this to TS but I'll plan to also handle this in a follow-up along with the Price block if that's ok.~~

Actually, I've changed my mind. I'm going to convert now because I will add a types interface for the context that we'll be able to use on the rest of the blocks.

Going forward, however, I think it will be best to approach the TS conversion in separate PRs due to the potential "rabbit-hole" of necessary adjustments to other parts of the code base.

danielwrobert committed:
Add PQ Context interface to shared type defs.

danielwrobert committed:
Convert all block JS files to TS.

danielwrobert committed:
Remove commented import from block file.

sunyatasattva commented:
Hello @danielwrobert! Thanks for the work you're making here.

I'm failing to properly test this PR, I'm not sure what I'm doing wrong.

1. I can't see the Product Rating in the inserter when I try to add it manually to my Product Query block.
2. If I follow your test steps, it appears on the editor but it asks me to choose a product. Even when I do that, it doesn't show up anymore.

![Screenshot 2022-10-12 at 06 26 04](https://user-images.githubusercontent.com/1847066/195249859-5686c5f4-a9cd-4251-8a8f-1e616782993a.png)

Is there something obvious that I might be missing?

FWIW I feel the test instructions should also include adding the block via the inserter. Also:

> Note that this block has a different behavior when it is used in the Product Query block as respect when it is used in All Products. We will address this issue in the next iteration https://github.com/woocommerce/woocommerce-blocks/issues/6949.

Seem to refer to the Add to Cart block. Is there a different behavior for this block as well?

danielwrobert committed:
Add typecasting to block function params.

As a workaround, added a general Record type but left a TODO to revisit
the proper object, as there is a mismatch in the shape of the default
object property types and the actual types.

sunyatasattva reviewed (changes requested):
Marking this as needing changes as we discussed in our call. Feel free to re-ask the review when ready.

danielwrobert committed:
Update inserter behavior.

Allows for the ability to add the rating block from in the inserter
(as long as it's an inner block of the listed parents in the config).
Also disables the placeholder product selector from being rendered
unnecessarily (i.e., when the context ID is present).

danielwrobert commented:
@sunyatasattva I've sorted out the issues with the inserter and the product selection interface that we discussed in 76ce6f6. I've also updated the test plan accordingly.

While looking into the block config, I noticed that the only difference between the [parent property in the shared config](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/atomic/blocks/product-elements/shared/config.tsx#L31) and the [parent property in these updates](https://github.com/woocommerce/woocommerce-blocks/pull/7352/files#diff-a1124248b7c49afbad741e61af47756aa717955fe8180c77a825b01da9fbaf06R25-R29) is the addition of the `core/post-template` block. Would it make sense to update the shared config now, since most (if not all) of these blocks will need that addition? Any exceptions can be overridden on a per-block basis.

What do you think? (cc @gigitux)

gigitux commented:
> @sunyatasattva I've sorted out the issues with the inserter and the product selection interface that we discussed in [76ce6f6](https://github.com/woocommerce/woocommerce-blocks/commit/76ce6f6334df1b7e2bc746b975d2ac4d104a190a). I've also updated the test plan accordingly.
> 
> While looking into the block config, I noticed that the only difference between the [parent property in the shared config](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/atomic/blocks/product-elements/shared/config.tsx#L31) and the [parent property in these updates](https://github.com/woocommerce/woocommerce-blocks/pull/7352/files#diff-a1124248b7c49afbad741e61af47756aa717955fe8180c77a825b01da9fbaf06R25-R29) is the addition of the `core/post-template` block. Would it make sense to update the shared config now, since most (if not all) of these blocks will need that addition? Any exceptions can be overridden on a per-block basis.
> 
> What do you think? (cc @gigitux)

Yes, it makes sense! Regarding this topic, I discovered a bug related to the `Add To Cart` block, and I updated the shared configuration. Happy to receive feedback! #7389

Aljullu commented on assets/js/atomic/blocks/product-elements/rating/edit.tsx:
```diff
@@ -3,22 +3,41 @@
 	return (
 		<div { ...blockProps }>
-			<Block { ...attributes } />
+			<Block { ...blockAttrs } />
```
I know this isn't specific from this PR, but I thought I would leave a comment here anyway. :upside_down_face: Would it be possible to pass `isDescendentOfQueryLoop` as a prop to `Block` so we don't need to save it in the block attributes?

gigitux replied:
Thanks for leaving the comment! 
No, it isn't possible because this flag is [important for the save function](https://github.com/woocommerce/woocommerce-blocks/blob/8411854c52bbd5b0a97f19016c468cb5ad9532e8/assets/js/atomic/blocks/product-elements/save.js/#L6-L14)

Reading the attributes, when `isDescendentOfQueryLoop` is true we don't save anything, in this way on the PHP side, we know when to do server-side rendering or not https://github.com/woocommerce/woocommerce-blocks/blob/8411854c52bbd5b0a97f19016c468cb5ad9532e8/src/BlockTypes/ProductButton.php/#L77-L81

Aljullu replied:
Ah, right. In the PHP side, do you think we could invert the check, so instead of checking the content, we check for `$block->context['postId']` first? Similar to [what GB does](https://github.com/WordPress/gutenberg/blob/8f00bd0b129ad4afaef95b4f029787fcfc1c274c/packages/block-library/src/post-title/index.php#L18). So if there is a `postId`, we render the markup in PHP, if there isn't, we render the `content` as before.

Not sure if that would work (I just did some quick testing), but I feel it strange that we need to store `isDescendentOfQueryLoop` in the block attributes. :confused:

gigitux replied:
Keep in mind that the GB block that you linked has only the "server side rendering version". Also, `context['postId']` is always available, even if the block isn't a child of the Product Query. The check that you linked is used for display or not the block if the post exists or not. We do the [same thing](https://github.com/woocommerce/woocommerce-blocks/blob/f8a6bab2ed31f1a0c4c0038c087dc20b33bd40a1/src/BlockTypes/ProductImage.php/#L196-L212).

We could use `queryId` instead of `postId` 🤔 

The problem is that we will save in `post_content` some HTML that we will not use... with the current approach we save nothing.

Aljullu replied:
Oooh, I see. Thanks for explaining. I think it's unfortunate that we need to save `isDescendentOfQueryLoop` as an attribute, but from what you say, it looks like there is no way we can get rid of it if we want the block to continue working in the context of the All Products block. :disappointed: 

I guess that if we migrate the All Products block to use the new Product Query block in the future, we will then be able to remove the `isDescendentOfQueryLoop` attribute.

gigitux replied:
> Oooh, I see. Thanks for explaining.

You're welcome! 💪 

> I guess that if we migrate the All Products block to use the new Product Query block in the future, we will then be able to remove the isDescendentOfQueryLoop attribute.

Yeah! Exactly!

danielwrobert committed:
Update parent inner blocks config.

Reassign parent array to ancestor array which allows for blocks to be
included with more flexibility - i.e., added within groups that are
children of the ancestor block.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/rating/block.tsx:
```diff
@@ -14,20 +13,49 @@ import {
+};
+
+const getAverageRating = (
+	// TODO: This is a workaround until we can confirm the correct type from ProductResponseItem.
```
Why can't we confirm the correct type?

Also note that we have an automation that creates new issues with every `TODO` comment. I'm not sure we want to create an issue for this specifically. Maybe we can create the issue that ties into some other TS migration?

danielwrobert replied:
> Why can't we confirm the correct type?

There is a mismatch in the data between the `ProductResponseItem` interface ([here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/types/type-defs/product-response.ts#L76-L77)) and the default data in `defaultProductData` ([here](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/shared/context/product-data-context.js#L34-L35)).

We want to confirm, specifically, if the `average_rating` property is actually a number or a string. I discussed this on a pairing chat with @gigitux yesterday and it seems that it would be better to move forward with this workaround for the time being, as opposed to holding up the block - we can then circle back to address this issue with some of the other TS issues / migrations in a follow-up task.

sunyatasattva replied:
I did a quick API call to confirm. `average_rating` is indeed a string. This is the response I got:

```
{
	"id": 16,
	"name": "Beanie",
	"parent": 0,
	"type": "simple",
	"variation": "",
	"permalink": "http:\/\/kirigami.local\/product\/beanie\/",
	"sku": "woo-beanie",
	"short_description": "<p>This is a simple product.<\/p>",
	"description": "<p>Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Vestibulum tortor quam, feugiat vitae, ultricies eget, tempor sit amet, ante. Donec eu libero sit amet quam egestas semper. Aenean ultricies mi vitae est. Mauris placerat eleifend leo.<\/p>",
	"on_sale": true,
	"prices": {
		"price": "1800",
		"regular_price": "2000",
		"sale_price": "1800",
		"price_range": null,
		"currency_code": "EUR",
		"currency_symbol": "€",
		"currency_minor_unit": 2,
		"currency_decimal_separator": ",",
		"currency_thousand_separator": ".",
		"currency_prefix": "",
		"currency_suffix": " €"
	},
	"price_html": "<del aria-hidden=\"true\"><span class=\"woocommerce-Price-amount amount\">20,00&nbsp;<span class=\"woocommerce-Price-currencySymbol\">&euro;<\/span><\/span><\/del> <ins><span class=\"woocommerce-Price-amount amount\">18,00&nbsp;<span class=\"woocommerce-Price-currencySymbol\">&euro;<\/span><\/span><\/ins>",
	"average_rating": "5.00",
	"review_count": 1,
	"images": [
		{
			"id": 45,
			"src": "http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2.jpg",
			"thumbnail": "http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-450x450.jpg",
			"srcset": "http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2.jpg 801w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-450x450.jpg 450w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-100x100.jpg 100w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-600x600.jpg 600w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-300x300.jpg 300w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-150x150.jpg 150w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-768x768.jpg 768w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-324x324.jpg 324w, http:\/\/kirigami.local\/wp-content\/uploads\/2022\/01\/beanie-2-416x416.jpg 416w",
			"sizes": "(max-width: 801px) 100vw, 801px",
			"name": "beanie-2.jpg",
			"alt": ""
		}
	],
	"categories": [
		{
			"id": 19,
			"name": "Accessories",
			"slug": "accessories",
			"link": "http:\/\/kirigami.local\/product-category\/clothing\/accessories\/"
		}
	],
	"tags": [],
	"attributes": [
		{
			"id": 1,
			"name": "Color",
			"taxonomy": "pa_color",
			"has_variations": false,
			"terms": [
				{
					"id": 24,
					"name": "Red",
					"slug": "red"
				}
			]
		}
	],
	"variations": [],
	"has_options": false,
	"is_purchasable": true,
	"is_in_stock": true,
	"is_on_backorder": false,
	"low_stock_remaining": null,
	"sold_individually": false,
	"add_to_cart": {
		"text": "Add to cart",
		"description": "Add &ldquo;Beanie&rdquo; to your cart",
		"url": "?add-to-cart=16",
		"minimum": 1,
		"maximum": 9999,
		"multiple_of": 1
	},
	"extensions": {}
}
```

So, my suggestion is to fix the wrong interface. I did a quick search on our repo, and it seems that the `average_rating` as a number is used in only two cases, so that could also be fixed relatively quickly.

If you still believe to be best not to fix the interface right now, you can do this:

```
product: Omit< ProductResponseItem, 'average_rating' > & {
  average_rating: string;
}
```

danielwrobert replied:
Thanks for the info here, @sunyatasattva, this is very helpful!

> it seems that the `average_rating` as a number is used in only two cases, so that could also be fixed relatively quickly.

Yeah, I am seeing the same thing - showing up in previews (`assets/js/previews/cart.ts` and `assets/js/previews/product.js`). While it should be a relatively quick fix, I think I'd prefer to address that separately as a follow-up, if that's ok. That way I can keep this work moving forward and scoped to the task. We can also test to confirm there's no impact on updating that interface without holding things up here.

danielwrobert replied:
I've created a new issue to address the incorrect type on the `ProductResponseItem` at #7408, and assigned myself to it.

I've also updated the type-casting here, per your suggestion, and removed the `TODO` comments in 9eab7cbc7.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/rating/edit.tsx:
```diff
@@ -3,22 +3,41 @@
-const Edit = ( { attributes } ) => {
+interface Props {
+	attributes: Attributes;
+	setAttributes: ( attributes: Record< string, unknown > ) => void;
```
🤔  Shouldn't `attributes` be of the type `Attributes` above?

It seems to me that these interface might be lacking all the other props, however… you can use the handy `BlockEditProps< T >` already available in the types.

danielwrobert replied:
Addressed in ad2b4decd.

sunyatasattva commented on assets/js/atomic/blocks/product-elements/rating/types.ts:
```diff
@@ -0,0 +1,3 @@
+export interface Attributes {
+	isDescendentOfQueryLoop: boolean;
+}
```
Shouldn't `productId?` also be available here?

danielwrobert replied:
Nice catch - added in ef2851590.

sunyatasattva commented on assets/js/types/type-defs/contexts.ts:
```diff
@@ -13,6 +13,12 @@ export type ValidationData = {
 	) => void;
 };
 
+export type ProductQueryContext = {
```
Is there a reason why this is a `type` as opposed to `interface` like the others you have created?

danielwrobert replied:
Addressed in ad2b4decd.

sunyatasattva commented on assets/js/types/type-defs/contexts.ts:
```diff
@@ -13,6 +13,12 @@ export type ValidationData = {
 };
 
+export type ProductQueryContext = {
+	query: object;
```
In general, `object` should never be used as a type. TSLint used to report it by default, and I just noticed [it doesn't anymore since last year](https://github.com/woocommerce/woocommerce-blocks/pull/7366).

In any case, `object` indicates also arrays, for instance. In general, you can use the utility type `Record` for your object needs.

In this case, we do have the `query` type: you can use `QueryBlockQuery & ProductQueryArguments`.

danielwrobert replied:
Addressed in ad2b4decd.

Thanks, this was very helpful! 🙂

sunyatasattva commented on assets/js/types/type-defs/contexts.ts:
```diff
@@ -13,6 +13,12 @@ export type ValidationData = {
 
+export type ProductQueryContext = {
+	query: object;
+	queryId: number;
+	productId: number;
```
Are those always available in the context or sometimes undefined?

danielwrobert replied:
Addressed in ad2b4decd.

Note: we don't need the `productId`, as it is passed in via the `attributes`. So I've removed it in that commit as well.

sunyatasattva reviewed (changes requested):
I've got some small change requests, but the work looks great! 🙌

danielwrobert committed:
Add productID to rating Attributes interface.

danielwrobert committed:
TS type casting and import adustments.

Some adjustments to utilize types that we already have available, along
with some syntax adjustments and more sensible import tweaks.

danielwrobert commented on assets/js/atomic/blocks/product-elements/rating/edit.tsx:
```diff
@@ -3,22 +3,40 @@
 import withProductSelector from '../shared/with-product-selector';
 import { BLOCK_TITLE, BLOCK_ICON } from './constants';
+import { Attributes } from './types';
+import { ProductQueryContext as Context } from '../../../../blocks/product-query/types';
```
Just noting that I realize this is not an ideal import approach by using the relative path. However, moving this to an alias includes updating/cleaning up the tsconfig(s) and webpack config - it would also be good to test any potential impact that might raise.

In the interest of keeping this PR scoped to supporting the PQ block, I propose addressing this in a follow-up item.

sunyatasattva replied:
What potential impacts do you see here? I agree with you on keeping the PR scoped, but also to not unnecessarily add any other small tasks which could be solved with little effort.

I see that perhaps you mean that if we introduce a `@woocommerce/blocks` alias path and migrate all our paths it could cause some unintended consequences, is that correct? While I'm not seeing too many unintended consequences for that either, we can add the alias and use it only here (and from now on) to keep this PR correctly scoped.

It's, in a sense, the same principle as not necessarily removing all our TS errors, but not add any further TS error. So here, we don't have to fix all relative paths, just not add wood to the fire, so to speak. And then, like you say, make another refactoring issue and PR to fix the older relative paths (or fix them as they come into play).

So, these are my 2c. But I'm approving the PR because it's such a small thing and if you disagree with it, I'm also ok with that.

danielwrobert committed:
Update type-casting to use ProductResponseItem

Instead of using the generic Record, we can utilize the
ProductResponseItem interface and set an omission for the average_rating
property until that is corrected to properly reflect the API response.

sunyatasattva reviewed (approved):
Still have a small comment about the relative path, but it's not mandatory if you don't want to fix it in this PR, so I am approving.

danielwrobert committed:
Add alias to blocks dir for imports.

Allows us to use exports from the blocks dir as "external" imports. This
way we do not need to write long, relative import paths (which can be
fragile in the long run).
